File: plugins/android-permissions/android/src/main/java/com/horamed/androidpermissions/AndroidPermissionsPlugin.java
package com.horamed.androidpermissions;

import android.app.AlarmManager;
import android.content.Context;
import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.PowerManager;
import android.provider.Settings;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

@CapacitorPlugin(name = "AndroidPermissions")
public class AndroidPermissionsPlugin extends Plugin {

    @PluginMethod
    public void checkExactAlarmPermission(PluginCall call) {
        boolean granted = true;
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            AlarmManager alarmManager = (AlarmManager) getContext().getSystemService(Context.ALARM_SERVICE);
            granted = alarmManager.canScheduleExactAlarms();
        }
        
        JSObject ret = new JSObject();
        ret.put("granted", granted);
        call.resolve(ret);
    }

    @PluginMethod
    public void requestExactAlarmPermission(PluginCall call) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            try {
                Intent intent = new Intent(Settings.ACTION_REQUEST_SCHEDULE_EXACT_ALARM);
                intent.setData(Uri.parse("package:" + getContext().getPackageName()));
                getActivity().startActivity(intent);
                call.resolve();
            } catch (Exception e) {
                call.reject("Failed to open exact alarm settings", e);
            }
        } else {
            call.resolve(); // Not needed on Android < 12
        }
    }

    @PluginMethod
    public void checkBatteryOptimization(PluginCall call) {
        PowerManager powerManager = (PowerManager) getContext().getSystemService(Context.POWER_SERVICE);
        boolean optimized = true;
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            String packageName = getContext().getPackageName();
            optimized = !powerManager.isIgnoringBatteryOptimizations(packageName);
        }
        
        JSObject ret = new JSObject();
        ret.put("optimized", optimized);
        call.resolve(ret);
    }

    @PluginMethod
    public void requestIgnoreBatteryOptimization(PluginCall call) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            try {
                Intent intent = new Intent(Settings.ACTION_REQUEST_IGNORE_BATTERY_OPTIMIZATIONS);
                intent.setData(Uri.parse("package:" + getContext().getPackageName()));
                getActivity().startActivity(intent);
                call.resolve();
            } catch (Exception e) {
                // Fallback to general battery optimization settings
                try {
                    Intent fallbackIntent = new Intent(Settings.ACTION_IGNORE_BATTERY_OPTIMIZATION_SETTINGS);
                    getActivity().startActivity(fallbackIntent);
                    call.resolve();
                } catch (Exception ex) {
                    call.reject("Failed to open battery optimization settings", ex);
                }
            }
        } else {
            call.resolve(); // Not needed on Android < 6
        }
    }

    @PluginMethod
    public void getDeviceInfo(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("manufacturer", Build.MANUFACTURER);
        ret.put("model", Build.MODEL);
        ret.put("androidVersion", Build.VERSION.RELEASE);
        ret.put("sdkInt", Build.VERSION.SDK_INT);
        call.resolve(ret);
    }

    @PluginMethod
    public void isDeviceIdleMode(PluginCall call) {
        boolean idle = false;
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            PowerManager powerManager = (PowerManager) getContext().getSystemService(Context.POWER_SERVICE);
            idle = powerManager.isDeviceIdleMode();
        }
        
        JSObject ret = new JSObject();
        ret.put("idle", idle);
        call.resolve(ret);
    }

    @PluginMethod
    public void openAppSettings(PluginCall call) {
        try {
            Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
            intent.setData(Uri.parse("package:" + getContext().getPackageName()));
            getActivity().startActivity(intent);
            call.resolve();
        } catch (Exception e) {
            call.reject("Failed to open app settings", e);
        }
    }
}

File: plugins/android-permissions/ios/Plugin/AndroidPermissionsPlugin.swift
import Foundation
import Capacitor

/**
 * iOS implementation of AndroidPermissions plugin
 * This plugin is Android-only, so iOS implementation is a no-op
 */
@objc(AndroidPermissionsPlugin)
public class AndroidPermissionsPlugin: CAPPlugin {
    
    @objc func checkExactAlarmPermission(_ call: CAPPluginCall) {
        call.resolve(["granted": true])
    }
    
    @objc func requestExactAlarmPermission(_ call: CAPPluginCall) {
        call.resolve()
    }
    
    @objc func checkBatteryOptimization(_ call: CAPPluginCall) {
        call.resolve(["optimized": false])
    }
    
    @objc func requestIgnoreBatteryOptimization(_ call: CAPPluginCall) {
        call.resolve()
    }
    
    @objc func getDeviceInfo(_ call: CAPPluginCall) {
        call.resolve([
            "manufacturer": "Apple",
            "model": UIDevice.current.model,
            "androidVersion": 0,
            "sdkInt": 0
        ])
    }
    
    @objc func isDeviceIdleMode(_ call: CAPPluginCall) {
        call.resolve(["idle": false])
    }
    
    @objc func openAppSettings(_ call: CAPPluginCall) {
        if let url = URL(string: UIApplication.openSettingsURLString) {
            DispatchQueue.main.async {
                UIApplication.shared.open(url, options: [:], completionHandler: nil)
            }
        }
        call.resolve()
    }
}
